libmamba/src/specs/match_spec.cpp

```diff
@@ -502,32 +502,58 @@ namespace mamba::specs
 
     auto MatchSpec::parse(std::string_view str) -> expected_parse_t<MatchSpec>
     {
-        auto parse_error = [&str](std::string_view err) -> tl::unexpected<ParseError>
+        std::string raw_match_spec_str = std::string(str);
+        raw_match_spec_str = util::strip(raw_match_spec_str);
+
+        // Remove any with space after binary operators, such as:
+        //  - `openmpi-4.1.4-ha1ae619_102`'s improperly encoded `constrains`: "cudatoolkit >= 10.2"
+        //  - `pytorch-1.13.0-cpu_py310h02c325b_0.conda`'s improperly encoded
+        //  `constrains`: "pytorch-cpu = 1.13.0", "pytorch-gpu = 99999999"
+        //  - `fipy-3.4.2.1-py310hff52083_3.tar.bz2`'s improperly encoded `constrains` or
+        //  `dep`: ">=4.5.2"
+        //  - `infokonoha-4.6.3-pyhd8ed1ab_0.tar.bz2`'s `kytea >=0.1.4, 0.2.0` -> `kytea
+        //  >=0.1.4,0.2.0`
+        // TODO: this solution reallocates memory several times potentially, but the
+        //  number of operators is small and the strings are short, so it must be fine.
+        //  If needed it can be optimized so that the string is only copied once.
+        for (const std::string& op : { ">=", "<=", "==", ">", "<", "!=", "=", "==", "," })
+        {
+            const std::string& bad_op = op + " ";
+            while (raw_match_spec_str.find(bad_op) != std::string::npos)
+            {
+                raw_match_spec_str = raw_match_spec_str.substr(0, raw_match_spec_str.find(bad_op)) + op
+                                     + raw_match_spec_str.substr(
+                                         raw_match_spec_str.find(bad_op) + bad_op.size()
+                                     );
+            }
+        }
+
```

Comment:
Maybe we won't have the choice and we will need to fix this this way, but I really think that we should do this properly and stop postponing everything to later, because it will just increase the complexity (making it harder to change things afterwards), especially regarding the `MatchSpec`...
IIRC this should be rather handled [here](https://github.com/mamba-org/mamba/blob/main/libmamba/src/specs/match_spec.cpp#L441-L488), so we need to adapt the logic accordingly (and try to keep the `string_view`).

Reply:
I agree with your proposal although there is some complexity with the low level parser: if we want to handle all the current cases while excluding the [PEP 508 environment markers](https://peps.python.org/pep-0508/#environment-markers), I am afraid that the amount of complexity to manage will be far more complex that this horrible yet working and short solution.

In my opinion, the long-term robust solution (as discussed in the past) is to define a grammar for `MatchSpec` and use lexers to generate parsers in applications. But this goes far beyond the scope of this PR or the time we have at hand.